src/inmanta/deploy/state.py

```diff
@@ -16,40 +16,31 @@
     Contact: code@inmanta.com
 """
 
+import contextlib
 import dataclasses
 import enum
 from collections import defaultdict
 from collections.abc import Mapping, Set
 from dataclasses import dataclass
 from enum import StrEnum
-from typing import Collection, TypeAlias
 
-from inmanta.agent import executor
 from inmanta.data.model import ResourceIdStr, ResourceType
 from inmanta.resources import Id
-from inmanta.util.collections import BidirectionalManyToManyMapping
+from inmanta.util.collections import BidirectionalManyMapping
 
 
-class RequiresProvidesMapping(BidirectionalManyToManyMapping[ResourceIdStr, ResourceIdStr]):
+class RequiresProvidesMapping(BidirectionalManyMapping[ResourceIdStr, ResourceIdStr]):
     def requires_view(self) -> Mapping[ResourceIdStr, Set[ResourceIdStr]]:
         return self
 
     def provides_view(self) -> Mapping[ResourceIdStr, Set[ResourceIdStr]]:
         return self.reverse_mapping()
 
 
-AttributeHash: TypeAlias = str
-
-
-class ResourceDetails(executor.ResourceDetails):
-
-    # FIXME: flatten out inheritance?
-
-    def __init__(self, id: ResourceIdStr, version: int, attributes: dict[str, object], attribute_hash: AttributeHash) -> None:
-        super().__init__(id, version, attributes)
-        self.attribute_hash = attribute_hash
-
-    attribute_hash: AttributeHash
+@dataclass(frozen=True)
+class ResourceDetails:
```

Comment:
This is one of the bigger changes I made, as proposed on Slack. I felt that the inheritance from `executor.ResourceDetails` was not appropriate here, because `executor.ResourceDetails` contains the full specification to deploy a resource, while this dataclass here is simply meant to express the current intent of a resource.

Most importantly imo, it should be version agnostic: the scheduler knows which version it reflects, and it makes sure that its model state reflects that version. If a new version is read from the database, unchanged resources should not be affected. The resource should become versioned only at the point where we actually commit to a version. Therefore I moved the construction of `executor.ResourceDetails` to the `tasks` module.

Reply:
I have to confess I'm always on the fence for these types of things. Do you duplicate the id in the object, adding an invariant that it matches the key in the mapping, or do you keep it contained to just the "data"? I'm not particularly attached to this, so I can add the id.